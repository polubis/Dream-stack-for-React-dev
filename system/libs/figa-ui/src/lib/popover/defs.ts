import type { UseToggleReturn } from '@system/figa-hooks';
import type { ReactNode } from 'react';
import type { SpacingKey } from '../theme-provider';

interface PopoverProps {
  className?: string;
  initialOpen?: boolean;
  offsetY?: SpacingKey;
  children: (toggler: UseToggleReturn) => ReactNode;
  trigger: (toggler: UseToggleReturn) => ReactNode;
}

export type { PopoverProps };
