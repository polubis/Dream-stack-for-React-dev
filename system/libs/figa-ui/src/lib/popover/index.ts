export * from './defs';
export * from './popover';
