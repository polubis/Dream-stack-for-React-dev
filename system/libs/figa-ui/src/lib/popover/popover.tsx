import {
  createContext,
  useContext,
  useEffect,
  useId,
  useMemo,
  useState,
} from 'react';
import type {
  PopoverContentProps,
  PopoverContext,
  PopoverProps,
  PopoverTriggerProps,
} from './defs';
import { useIsomorphicLayoutEffect, usePortal } from '@system/figa-hooks';
import { spacing } from '../shared';
import { Box } from '../box';

const Context = createContext<PopoverContext | null>(null);

const Popover = ({
  closeMode = 'own',
  offsetX = 0,
  offsetY = 150,
  children,
  openOnInit = false,
}: PopoverProps) => {
  const id = useId();
  const [opened, setOpened] = useState(false);

  useEffect(() => {
    openOnInit && setOpened(true);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const value = useMemo(
    () => ({
      closeMode,
      offsetX,
      offsetY,
      opened,
      triggerId: `popover-trigger-${id}`,
      contentId: `popover-content-${id}`,
      closed: !opened,
      open: () => {
        setOpened(true);
      },
      close: () => {
        setOpened(false);
      },
      toggle: () => {
        setOpened((opened) => !opened);
      },
    }),
    [closeMode, offsetX, offsetY, opened, id]
  );

  return <Context.Provider value={value}>{children}</Context.Provider>;
};

const usePopover = () => {
  const ctx = useContext(Context);

  if (!ctx) throw Error('You need to wrap your components with <Popover>');

  return ctx;
};

const Trigger = ({ children }: PopoverTriggerProps) => {
  const { triggerId } = usePopover();

  return (
    <div id={triggerId} className="popover-trigger">
      {children}
    </div>
  );
};

const Content = ({ children, className, ...props }: PopoverContentProps) => {
  const { triggerId, contentId, close, closed, offsetY, offsetX, closeMode } =
    usePopover();
  const { render } = usePortal();

  useIsomorphicLayoutEffect(() => {
    if (closed) return;

    const trigger = document.getElementById(triggerId);
    const content = document.getElementById(contentId);

    if (!trigger || !content) throw Error('Cannot find Trigger or Content');

    const triggerRect = trigger.getBoundingClientRect();
    const contentRect = content.getBoundingClientRect();

    const isTriggerRight = triggerRect.left >= window.innerWidth / 2;
    const isTriggerBottom = triggerRect.top >= window.innerHeight / 2;

    content.style.left = `${
      isTriggerRight
        ? triggerRect.left - contentRect.width + triggerRect.width
        : triggerRect.left
    }px`;
    content.style.top = `${
      isTriggerBottom
        ? triggerRect.top - contentRect.height - spacing.parse(offsetY)
        : triggerRect.top + triggerRect.height + spacing.parse(offsetY)
    }px`;

    const isExceedingWindowWidth = contentRect.width > window.innerWidth;

    if (isExceedingWindowWidth) {
      content.style.width = '96%';
      content.style.maxWidth = 'unset';
      content.style.minWidth = 'unset';
      content.style.left = '2%';
    }
  }, [closed, offsetX, offsetY, triggerId, contentId]);

  if (closed) return null;

  if (closeMode === 'backdrop') {
    return render(
      <>
        <div className="backdrop popover-backdrop" onClick={close} />
        <Box {...props} id={contentId} className="popover-content">
          {children}
        </Box>
      </>
    );
  }

  return render(
    <Box {...props} id={contentId} className="popover-content">
      {children}
    </Box>
  );
};

Popover.Trigger = Trigger;
Popover.Content = Content;
Popover.use = usePopover;

export { Popover };
