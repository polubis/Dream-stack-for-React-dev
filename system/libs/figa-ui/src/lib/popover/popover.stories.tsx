import type { Story, Meta } from '@storybook/react';

import { Popover } from './popover';
import { Button } from '../button';
import { Font } from '../font';
import { CloseIcon } from '../icon';
import { Box, BoxProps } from '../box';

export default {
  component: Popover,
  title: 'Popover',
  parameters: {
    viewport: {
      viewports: {
        tinyMobile: {
          name: 'tinyMobile',
          styles: {
            width: '300px',
            height: '800px',
          },
        },
      },
    },
  },
} as Meta;

const Trigger = () => {
  const { toggle, opened } = Popover.use();

  return (
    <Popover.Trigger>
      <Button onClick={toggle}>{opened ? 'Close' : 'Open'}</Button>
    </Popover.Trigger>
  );
};

const Content = (props: BoxProps) => {
  const { close } = Popover.use();

  return (
    <Popover.Content
      padding={[250, 250, 250, 250]}
      orientation="row"
      variant="outlined"
      minWidth="300px"
      maxWidth="250px"
      {...props}
    >
      {props.children ?? (
        <>
          <Font variant="h5">My content is here</Font>
          <Button onClick={close}>
            <CloseIcon />
          </Button>
        </>
      )}
    </Popover.Content>
  );
};

const DefaultTemplate: Story = () => {
  return (
    <Box between style={{ minHeight: '100vh' }} padding={[300, 300, 300, 300]}>
      <Box orientation="row" between>
        <Popover openOnInit>
          <Trigger />
          <Content />
        </Popover>
        <Popover openOnInit>
          <Trigger />
          <Content />
        </Popover>
        <Popover openOnInit>
          <Trigger />
          <Content />
        </Popover>
      </Box>
      <Box orientation="row" between>
        <Popover openOnInit>
          <Trigger />
          <Content />
        </Popover>
        <Popover openOnInit>
          <Trigger />
          <Content />
        </Popover>
        <Popover openOnInit>
          <Trigger />
          <Content />
        </Popover>
      </Box>
      <Box orientation="row" between>
        <Popover openOnInit>
          <Trigger />
          <Content />
        </Popover>
        <Popover openOnInit>
          <Trigger />
          <Content />
        </Popover>
        <Popover openOnInit>
          <Trigger />
          <Content />
        </Popover>
      </Box>
    </Box>
  );
};

const WithTooBigContentInXAxis: Story = () => {
  return (
    <Box padding={[300, 300, 300, 300]}>
      <Box spacing={[2000]}>
        <Popover openOnInit>
          <Trigger />
          <Content maxWidth={undefined} minWidth="1600px" />
        </Popover>
      </Box>
      <Box right>
        <Popover openOnInit>
          <Trigger />
          <Content maxWidth={undefined} minWidth="1600px" />
        </Popover>
      </Box>
    </Box>
  );
};

const WithCloseModeBackdrop: Story = () => {
  return (
    <Box padding={[300, 300, 300, 300]}>
      <Font variant="h6">
        s simply dummy text of the printing and typesetting industry. Lorem
        Ipsum has been the industry's standard dummy text ever since the 1500s,
        when an unknown printer took a galley of type and scrambled it to make a
        type specimen book. It has survived not only five centuries, but also
        the leap into electronic typesetting, remaining essentially unchanged.
        It was popularised in the 1960s with the release of Letraset sheets
        containing Lorem Ipsum passages, and more recently with desktop
        publishing software like Aldus PageMaker including versions of Lorem
        Ipsum.
      </Font>
      <Popover openOnInit closeMode="backdrop">
        <Trigger />
        <Content />
      </Popover>
    </Box>
  );
};

const WithCustomScrollInsideTemplate: Story = () => {
  return (
    <Box padding={[300, 300, 300, 300]}>
      <Popover openOnInit>
        <Trigger />
        <Content maxWidth={undefined} minWidth={undefined}>
          <div
            style={{ maxHeight: '400px', maxWidth: '300px', overflowY: 'auto' }}
          >
            <Font variant="h1">
              s simply dummy text of the printing and typesetting industry.
              Lorem Ipsum has been the industry's standard dummy text ever since
              the 1500s, when an unknown printer took a galley of type and
              scrambled it to make a type specimen book. It has survived not
              only five centuries, but also the leap into electronic
              typesetting, remaining essentially unchanged. It was popularised
              in the 1960s with the release of Letraset sheets containing Lorem
              Ipsum passages, and more recently with desktop publishing software
              like Aldus PageMaker including versions of Lorem Ipsum.
            </Font>
          </div>
        </Content>
      </Popover>
    </Box>
  );
};

export const Default = DefaultTemplate.bind({});
Default.args = {};

export const WithTooBigContentOnXAxis = WithTooBigContentInXAxis.bind({});
WithTooBigContentOnXAxis.args = {};

export const WithBackdropCloseMode = WithCloseModeBackdrop.bind({});
WithBackdropCloseMode.args = {};

export const WithCustomScrollInside = WithCustomScrollInsideTemplate.bind({});
WithCustomScrollInside.args = {};

export const OnMobile = DefaultTemplate.bind({});
OnMobile.args = {};
OnMobile.parameters = {
  viewport: {
    defaultViewport: 'tinyMobile',
  },
  chromatic: { delay: 1000 },
};
