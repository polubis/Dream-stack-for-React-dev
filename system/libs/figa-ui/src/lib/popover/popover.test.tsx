import { fireEvent, render, screen } from '@testing-library/react';
import { Popover } from './popover';

describe('Popover can be used when', () => {
  it('[FRAGILE] assigns classes', () => {
    const { container, asFragment } = render(
      <Popover className="my-class" trigger={() => <button>Click</button>}>
        {() => <div>Content</div>}
      </Popover>
    );

    const component = container.querySelector('.popover');

    expect(component?.className).toContain('popover my-class');
    expect(asFragment()).toMatchSnapshot();
  });

  it('allows to open content initially', () => {
    render(
      <Popover initialOpen trigger={() => <button>Click</button>}>
        {() => <div>Content</div>}
      </Popover>
    );

    screen.getByText(/Content/);
  });

  it('opens content', () => {
    render(
      <Popover
        trigger={({ toggle }) => <button onClick={toggle}>Click</button>}
      >
        {({ toggle }) => <div onClick={toggle}>Content</div>}
      </Popover>
    );

    fireEvent.click(screen.getByText(/Click/));

    screen.getByText(/Content/);
  });

  it('closes content', () => {
    render(
      <Popover initialOpen trigger={() => <button>Click</button>}>
        {({ toggle }) => <div onClick={toggle}>Content</div>}
      </Popover>
    );

    fireEvent.click(screen.getByText(/Content/));

    expect(screen.queryByText(/Content/)).toBeFalsy();
  });

  it('[FRAGILE] shifts popover content to left/right and top/bottom', () => {
    const initialWidth = window.innerWidth;
    const initialHeight = window.innerHeight;

    window.innerWidth = 900;
    window.innerHeight = 800;

    const { asFragment } = render(
      <Popover
        trigger={({ toggle }) => (
          <button style={{ width: '150px', height: '100px' }} onClick={toggle}>
            Click
          </button>
        )}
      >
        {({ toggle }) => (
          <div
            style={{
              height: '200px',
              width: '200px',
            }}
            onClick={toggle}
          >
            Content
          </div>
        )}
      </Popover>
    );

    fireEvent.click(screen.getByText(/Click/));

    expect(asFragment()).toMatchSnapshot();

    window.innerWidth = initialWidth;
    window.innerHeight = initialHeight;
  });
});
