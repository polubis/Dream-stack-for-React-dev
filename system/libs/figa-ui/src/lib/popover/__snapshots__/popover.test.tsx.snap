// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover can be used when [FRAGILE] assigns classes 1`] = `
<DocumentFragment>
  <div
    class="popover my-class"
  >
    <div
      class="popover-trigger"
    >
      <button>
        Click
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Popover can be used when [FRAGILE] shifts popover content to left/right and top/bottom 1`] = `
<DocumentFragment>
  <div
    class="popover"
  >
    <div
      class="popover-trigger"
    >
      <button
        style="width: 150px; height: 100px;"
      >
        Click
      </button>
    </div>
    <div
      class="popover-content"
      style="right: -0px; top: 12px;"
    >
      <div
        style="height: 200px; width: 200px;"
      >
        Content
      </div>
    </div>
  </div>
</DocumentFragment>
`;
